[Feature] Heterogeneous Environments compatibility

matteobettini opened the pull request:
This introduces tests for het envs and key modifications to

- rollout
- step_mdp

vmoens commented on torchrl/data/utils.py:
```diff
@@ -42,6 +43,36 @@
+) -> TensorDictBase:
+    """Densely stack a list of TensorDictBase objects (or a LazyStackedTensorDict) given that they have the same structure.
+
+    This must be used when some of the tds involved can have LazyTds among keys (or keys of keys, recursively).
```
let's avoid abbreviations in docstrings.
Even better: we can link to tensordict classes by using 
```
:class:`tensordict.stuff`
```

vmoens commented on torchrl/data/utils.py:
```diff
@@ -42,6 +43,36 @@
+    In those cases, calling `torch.stack(td_list).to_tensordict()` is infeasible.
+    Thus, this function provides an alternative for densifying the list provided.
+
+    Args:
```
I think an example for this method would be helpful.

vmoens reviewed (commented):
Thanks for this, really useful

matteobettini commented on torchrl/data/utils.py:
```diff
@@ -42,6 +43,36 @@
 INDEX_TYPING = Union[None, int, slice, str, Tensor, List[Any], Tuple[Any, ...]]
 
 
+def dense_stack_tds(
```
I'll write some tests and examples for it.

I am down to move it to tensodict, I put it here because I thought we said we wanted the utis to be here for the deprecation move

matteobettini commented:
So i have updated and added all the requested changes.
my takes on the new function are

- it should be public because any user which has some nested lazy stacks somewhere should use it after calling `env.rollout(return_contiguous=False)`
- my personal preference is that it should be in tensordict utils but the choice is yours

vmoens commented:
> So i have updated and added all the requested changes. my takes on the new function are
> 
> * it should be public because any user which has some nested lazy stacks somewhere should use it after calling `env.rollout(return_contiguous=False)`
> * my personal preference is that it should be in tensordict utils but the choice is yours

let's put it in tensordict then!

vmoens commented on torchrl/envs/utils.py:
```diff
@@ -219,44 +219,63 @@ def _set_single_key(source, dest, key, clone=False):
-            new_val = dest.get(key, None)
-            if new_val is None:
-                new_val = val.empty()
+        shape = source.get_item_shape(key)
+        if -1 not in shape:
```
what's the overhead here?
Since this will likely be very frequent I'm a bit worried by increasing the control flow for every key

vmoens replied:
For a regular td, this op means that we'll be accessing the tensor twice through get (once to get its shape, a second time to get it). Accessing it once should suffice.
Not I understand that we can't use `get` if the tensor is heterogeneous... A try/except may not be the right solution either since it'll be slow for heterogeneous envs. However regular envs won't be impacted

matteobettini replied:
I am not sure what I can do here
- in the past i was using try blocks, but you made me move away from them
- an isinstance wont work since the key might be het even if the container is a regular td

so I can move back to try except if you prefer

matteobettini replied:
lemme switch to try blocks

ideally tho this problem could be solved once and for all in tensordict, with an efficient way to understand if a key is het

vmoens replied:
yep that's what I'm thinking too
If we have tensorstack we can even leave the block with try / except and it won't break (I guess)

vmoens replied:
Just make sure you comment the try / excpet saying that it's temporary and for efficiency purpose

matteobettini commented:
updated

vmoens reviewed (approved):
_**Het**_ groene licht is gekomen
